qubesbuilder/plugins/fetch/scripts/verify-file

```diff
@@ -0,0 +1,132 @@
+    --output-dir                Output directory
+"
+}
+
+if ! OPTS=$(getopt -o hf:c:m:s:p:o: --long help,untrusted-file:,checksum-file:,checksum-cmd:,untrusted-signature-file:,pubkey-file:,output-dir: -n "$0" -- "$@"); then
+    echo "ERROR: Failed while parsing options."
+    exit 1
+fi
+
+eval set -- "$OPTS"
+
+PUBKEY_FILE=()
+
+while [[ $# -gt 0 ]]; do
+    case "$1" in
+        -h | --help) usage ;;
+        -f | --untrusted-file ) UNTRUSTED_FILE="$2"; shift ;;
+        -c | --checksum-file ) FILE_CHECKSUM="$2"; shift ;;
+        -m | --checksum-cmd ) CHECKSUM_CMD="$2"; shift ;;
+        -s | --untrusted-signature-file ) UNTRUSTED_SIGNATURE_FILE="$2"; shift ;;
+        -p | --pubkey-file ) PUBKEY_FILE+=("$2"); shift ;;
+        -o | --output-dir ) OUTPUT_DIR="$2"; shift ;;
+    esac
+    shift
+done
+
+if [ -z "${UNTRUSTED_FILE}" ]; then
+    echo "ERROR: Please provide UNTRUSTED_FILE."
+    exit 1
+fi
+
+if [ -n "${OUTPUT_DIR}" ]; then
+    mkdir -p "${OUTPUT_DIR}"
+else
+    OUTPUT_DIR='.'
+fi
+
+UNTRUSTED_FILE_NAME="$(basename "${UNTRUSTED_FILE}")"
+FILE_NAME="${UNTRUSTED_FILE_NAME#untrusted_}"
+
+if [ -z "${FILE_CHECKSUM}" ] && [ -z "${UNTRUSTED_SIGNATURE_FILE}" ] && [ -z "${PUBKEY_FILE[*]}" ]; then
+    echo "ERROR: Please provide either CHECKSUM or, UNTRUSTED_SIGNATURE_FILE and PUBKEY_FILE."
+    exit 1
+fi
+
+if [ -n "${FILE_CHECKSUM}" ] && ! [ -e "${FILE_CHECKSUM}" ]; then
+    echo "ERROR: Cannot find '${FILE_CHECKSUM}' checksum for '${FILE_NAME}'."
+    exit 1
+fi
+
+if [ -n "${UNTRUSTED_SIGNATURE_FILE}" ] && [ -z "${PUBKEY_FILE}" ]; then
+    echo "ERROR: Please provide PUBKEY_FILE to be used with UNTRUSTED_SIGNATURE_FILE."
+    exit 1
+fi
+
+if [ -z "${UNTRUSTED_SIGNATURE_FILE}" ] && [ -n "${PUBKEY_FILE}" ]; then
+    echo "ERROR: Please provide UNTRUSTED_SIGNATURE_FILE to be used with PUBKEY_FILE."
+    exit 1
+fi
+
+if [ -n "${UNTRUSTED_SIGNATURE_FILE}" ] && [ ! -e "${PUBKEY_FILE}" ]; then
+    echo "ERROR: Cannot find '${PUBKEY_FILE}' pubkey for '${UNTRUSTED_SIGNATURE_FILE}'."
+    exit 1
+fi
+
+if [ -n "${FILE_CHECKSUM}" ]; then
+    if [ -z "${CHECKSUM_CMD}" ]; then
+        CHECKSUM_CMD=sha256sum
+    fi
+
+    # Check downloaded file with respect to provided checksum file
+    "${CHECKSUM_CMD}" --status -c <(printf "%s  -\n" "$(cat "${FILE_CHECKSUM}")") \
+      < "${UNTRUSTED_FILE}" || { echo "${CHECKSUM_CMD}: wrong checksum on '${FILE_NAME}'."; exit 1; }
+
+elif [ -n "${UNTRUSTED_SIGNATURE_FILE}" ] && [ -n "${PUBKEY_FILE}" ]; then
+    UNTRUSTED_SIGNATURE_FILE_NAME="$(basename "$UNTRUSTED_SIGNATURE_FILE")"
+    SIGNATURE_FILE_NAME="${UNTRUSTED_SIGNATURE_FILE_NAME#untrusted_}"
+
+    # Import pubkey
+    keyring="$(mktemp -u)"
+    for pubkey_file in "${PUBKEY_FILE[@]}"; do
+        gpg --no-auto-check-trustdb --no-default-keyring --keyring "$keyring" -q --import "${pubkey_file}"
+    done
+
+    # Verify file
+    gpgv --keyring "$keyring" "${UNTRUSTED_SIGNATURE_FILE}" "${UNTRUSTED_FILE}" 2>/dev/null
```

Comment:
I recommend using `sqv` instead of `gpgv`, unless you want to use the Python OpenPGP parser on the signature first.